# **Exploratory Data Analysis**:

Understanding the dataset to explore how the data is present in the database and if there is a need of creating some aggregated tables that can help with:

Vendor selection for profitability
Product Pricing Optimization


In [2]:
import pandas as pd
import mysql.connector
from sqlalchemy import create_engine

In [3]:
pip show mysql-connector-python

Name: mysql-connector-python
Version: 9.4.0
Summary: A self-contained Python driver for communicating with MySQL servers, using an API that is compliant with the Python Database API Specification v2.0 (PEP 249).
Home-page: https://dev.mysql.com/doc/connector-python/en/
Author: Oracle and/or its affiliates
Author-email: 
License: GNU GPLv2 (with FOSS License Exception)
Location: C:\Users\vrand\anaconda3\Lib\site-packages
Requires: 
Required-by: 
Note: you may need to restart the kernel to use updated packages.


In [4]:
pip install --upgrade mysql-connector-python

Note: you may need to restart the kernel to use updated packages.


In [5]:
host="localhost"
user="root"
password='0028'
database="suviddha_supermarket"
conn=mysql.connector.connect(host="localhost",
user="root",
password=r'0028',
database="suviddha_supermarket")

In [6]:
engine=create_engine("mysql+mysqlconnector://root:0028@localhost:3306/suviddha_supermarket")

In [7]:
query='select * from purchases'

In [8]:
df=pd.read_sql(query,engine)
df

InventoryId  Store  Brand                   Description  \
0           69_MOUNTMEND_8412     69   8412     Tequila Ocho Plata Fresno   
1            30_CULCHETH_5255     30   5255  TGI Fridays Ultimte Mudslide   
2           34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced   
3         1_HARDERSFIELD_5255      1   5255  TGI Fridays Ultimte Mudslide   
4           76_DONCASTER_2034     76   2034     Glendalough Double Barrel   
...                       ...    ...    ...                           ...   
2372469     49_GARIGILL_22298     49  22298       Zorvino Vyds Sangiovese   
2372470  1_HARDERSFIELD_19556      1  19556    Zorvino Bacca Z Blackberry   
2372471   66_EANVERNESS_22297     66  22297            Zorvino Vyds Pearz   
2372472    69_MOUNTMEND_19557     69  19557  Zorvino Fragole Z Strawberry   
2372473    55_DRY GULCH_22298     55  22298       Zorvino Vyds Sangiovese   

          Size  VendorNumber                   VendorName  PONumber  \
0        750mL           105  ALTAMAR BRANDS LLC               8124   
1        1.75L          4466  AMERICAN VINTAGE BEVERAGE        8137   
2        1.75L          4466  AMERICAN VINTAGE BEVERAGE        8137   
3        1.75L          4466  AMERICAN VINTAGE BEVERAGE        8137   
4        750mL           388  ATLANTIC IMPORTING COMPANY       8169   
...        ...           ...                          ...       ...   
2372469  750mL         90058  ZORVINO VINEYARDS               13593   
2372470  750mL         90058  ZORVINO VINEYARDS               13593   
2372471  750mL         90058  ZORVINO VINEYARDS               13593   
2372472  750mL         90058  ZORVINO VINEYARDS               13593   
2372473  750mL         90058  ZORVINO VINEYARDS               13593   

             PODate ReceivingDate InvoiceDate     PayDate  PurchasePrice  \
0        2023-12-21    2024-01-02  2024-01-04  2024-02-16          35.71   
1        2023-12-22    2024-01-01  2024-01-07  2024-02-21           9.35   
2        2023-12-22    2024-01-02  2024-01-07  2024-02-21           9.41   
3        2023-12-22    2024-01-01  2024-01-07  2024-02-21           9.35   
4        2023-12-24    2024-01-02  2024-01-09  2024-02-16          21.32   
...             ...           ...         ...         ...            ...   
2372469  2024-12-19    2024-12-28  2025-01-09  2025-02-06           8.06   
2372470  2024-12-19    2024-12-27  2025-01-09  2025-02-06           9.39   
2372471  2024-12-19    2024-12-26  2025-01-09  2025-02-06           6.75   
2372472  2024-12-19    2024-12-26  2025-01-09  2025-02-06           9.39   
2372473  2024-12-19    2024-12-28  2025-01-09  2025-02-06           8.06   

         Quantity  Dollars  Classification  
0               6   214.26               1  
1               4    37.40               1  
2               5    47.05               1  
3               6    56.10               1  
4               5   106.60               1  
...           ...      ...             ...  
2372469        12    96.72               2  
2372470        12   112.68               2  
2372471        12    81.00               2  
2372472        12   112.68               2  
2372473        12    96.72               2  

[2372474 rows x 16 columns]

In [9]:
tables = pd.read_sql_query("SHOW TABLES", conn)
tables

C:\Users\vrand\AppData\Local\Temp\ipykernel_27676\2393634187.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  tables = pd.read_sql_query("SHOW TABLES", conn)


Tables_in_suviddha_supermarket
0                begin_inventory
1                  end_inventory
2                purchase_prices
3                      purchases
4                          sales
5                 vendor_invoice

In [10]:
for table in tables.iloc[:, 0]:  # getting actual table names
    print('-' * 50, table, '-' * 50)
    print('count of records:', pd.read_sql(f"SELECT COUNT(*) AS count FROM `{table}`", engine)['count'].values[0])
    display(pd.read_sql(f'select * from {table} limit 5',engine))

-------------------------------------------------- begin_inventory --------------------------------------------------
count of records: 5543


VendorNumber                   VendorName InvoiceDate  PONumber  \
0           105  ALTAMAR BRANDS LLC           2024-01-04      8124   
1          4466  AMERICAN VINTAGE BEVERAGE    2024-01-07      8137   
2           388  ATLANTIC IMPORTING COMPANY   2024-01-09      8169   
3           480  BACARDI USA INC              2024-01-12      8106   
4           516  BANFI PRODUCTS CORP          2024-01-07      8170   

       PODate     PayDate  Quantity    Dollars  Freight Approval  
0  2023-12-21  2024-02-16         6     214.26     3.47     None  
1  2023-12-22  2024-02-21        15     140.55     8.57     None  
2  2023-12-24  2024-02-16         5     106.60     4.61     None  
3  2023-12-20  2024-02-05     10100  137483.78  2935.20     None  
4  2023-12-24  2024-02-12      1935   15527.25   429.20     None

-------------------------------------------------- end_inventory --------------------------------------------------
count of records: 224489


InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   
2  1_HARDERSFIELD_63      1  HARDERSFIELD     63   Herradura Reposado Tequila   
3  1_HARDERSFIELD_72      1  HARDERSFIELD     72         No. 3 London Dry Gin   
4  1_HARDERSFIELD_75      1  HARDERSFIELD     75    Three Olives Tomato Vodka   

    Size  onHand  Price     endDate  
0  750mL      11  12.99  2024-12-31  
1  750mL       7  36.99  2024-12-31  
2  750mL       7  38.99  2024-12-31  
3  750mL       4  34.99  2024-12-31  
4  750mL       7  14.99  2024-12-31

-------------------------------------------------- purchase_prices --------------------------------------------------
count of records: 12261


Brand                  Description  Price   Size Volume  Classification  \
0     58  Gekkeikan Black & Gold Sake  12.99  750mL    750               1   
1     62     Herradura Silver Tequila  36.99  750mL    750               1   
2     63   Herradura Reposado Tequila  38.99  750mL    750               1   
3     72         No. 3 London Dry Gin  34.99  750mL    750               1   
4     75    Three Olives Tomato Vodka  14.99  750mL    750               1   

   PurchasePrice  VendorNumber                   VendorName  
0           9.28          8320  SHAW ROSS INT L IMP LTD      
1          28.67          1128  BROWN-FORMAN CORP            
2          30.46          1128  BROWN-FORMAN CORP            
3          26.11          9165  ULTRA BEVERAGE COMPANY LLP   
4          10.94          7245  PROXIMO SPIRITS INC.

-------------------------------------------------- purchases --------------------------------------------------
count of records: 2372474


InventoryId  Store  Brand                   Description   Size  \
0    69_MOUNTMEND_8412     69   8412     Tequila Ocho Plata Fresno  750mL   
1     30_CULCHETH_5255     30   5255  TGI Fridays Ultimte Mudslide  1.75L   
2    34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
3  1_HARDERSFIELD_5255      1   5255  TGI Fridays Ultimte Mudslide  1.75L   
4    76_DONCASTER_2034     76   2034     Glendalough Double Barrel  750mL   

   VendorNumber                   VendorName  PONumber      PODate  \
0           105  ALTAMAR BRANDS LLC               8124  2023-12-21   
1          4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
2          4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
3          4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
4           388  ATLANTIC IMPORTING COMPANY       8169  2023-12-24   

  ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0    2024-01-02  2024-01-04  2024-02-16          35.71         6   214.26   
1    2024-01-01  2024-01-07  2024-02-21           9.35         4    37.40   
2    2024-01-02  2024-01-07  2024-02-21           9.41         5    47.05   
3    2024-01-01  2024-01-07  2024-02-21           9.35         6    56.10   
4    2024-01-02  2024-01-09  2024-02-16          21.32         5   106.60   

   Classification  
0               1  
1               1  
2               1  
3               1  
4               1

-------------------------------------------------- sales --------------------------------------------------
count of records: 1048575


InventoryId Store Brand                 Description        Size  \
0  1_HARDERSFIELD_1004     1  1004  Jim Beam w/2 Rocks Glasses       750mL   
1  1_HARDERSFIELD_1004     1  1004  Jim Beam w/2 Rocks Glasses       750mL   
2  1_HARDERSFIELD_1004     1  1004  Jim Beam w/2 Rocks Glasses       750mL   
3  1_HARDERSFIELD_1004     1  1004  Jim Beam w/2 Rocks Glasses       750mL   
4  1_HARDERSFIELD_1005     1  1005     Maker's Mark Combo Pack  375mL 2 Pk   

  SalesQuantity SalesDollars SalesPrice   SalesDate Volume Classification  \
0             1        16.49      16.49  01-01-2024    750              1   
1             2        32.98      16.49  02-01-2024    750              1   
2             1        16.49      16.49  03-01-2024    750              1   
3             1        14.49      14.49  08-01-2024    750              1   
4             2        69.98      34.99  09-01-2024    375              1   

  ExciseTax VendorNo                   VendorName  
0      0.79    12546  JIM BEAM BRANDS COMPANY      
1      1.57    12546  JIM BEAM BRANDS COMPANY      
2      0.79    12546  JIM BEAM BRANDS COMPANY      
3      0.79    12546  JIM BEAM BRANDS COMPANY      
4      0.79    12546  JIM BEAM BRANDS COMPANY

-------------------------------------------------- vendor_invoice --------------------------------------------------
count of records: 5543


VendorNumber                   VendorName InvoiceDate  PONumber  \
0           105  ALTAMAR BRANDS LLC           2024-01-04      8124   
1          4466  AMERICAN VINTAGE BEVERAGE    2024-01-07      8137   
2           388  ATLANTIC IMPORTING COMPANY   2024-01-09      8169   
3           480  BACARDI USA INC              2024-01-12      8106   
4           516  BANFI PRODUCTS CORP          2024-01-07      8170   

       PODate     PayDate  Quantity    Dollars  Freight Approval  
0  2023-12-21  2024-02-16         6     214.26     3.47     None  
1  2023-12-22  2024-02-21        15     140.55     8.57     None  
2  2023-12-24  2024-02-16         5     106.60     4.61     None  
3  2023-12-20  2024-02-05     10100  137483.78  2935.20     None  
4  2023-12-24  2024-02-12      1935   15527.25   429.20     None

In [11]:
#Filtered out the rows which have vendornumber as 4466 in purchase

purchases=pd.read_sql_query("select * from purchases where vendornumber=4466",engine)
purchases

InventoryId  Store  Brand                   Description   Size  \
0        30_CULCHETH_5255     30   5255  TGI Fridays Ultimte Mudslide  1.75L   
1       34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
2     1_HARDERSFIELD_5255      1   5255  TGI Fridays Ultimte Mudslide  1.75L   
3       38_GOULCREST_5215     38   5215  TGI Fridays Long Island Iced  1.75L   
4     59_CLAETHORPES_5215     59   5215  TGI Fridays Long Island Iced  1.75L   
...                   ...    ...    ...                           ...    ...   
2187     81_PEMBROKE_5215     81   5215  TGI Fridays Long Island Iced  1.75L   
2188   62_KILMARNOCK_5255     62   5255  TGI Fridays Ultimte Mudslide  1.75L   
2189    34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
2190     6_GOULCREST_5215      6   5215  TGI Fridays Long Island Iced  1.75L   
2191    35_HALIVAARA_5255     35   5255  TGI Fridays Ultimte Mudslide  1.75L   

      VendorNumber                   VendorName  PONumber      PODate  \
0             4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
1             4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
2             4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
3             4466  AMERICAN VINTAGE BEVERAGE        8207  2023-12-27   
4             4466  AMERICAN VINTAGE BEVERAGE        8207  2023-12-27   
...            ...                          ...       ...         ...   
2187          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2188          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2189          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2190          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2191          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   

     ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0       2024-01-01  2024-01-07  2024-02-21           9.35         4    37.40   
1       2024-01-02  2024-01-07  2024-02-21           9.41         5    47.05   
2       2024-01-01  2024-01-07  2024-02-21           9.35         6    56.10   
3       2024-01-07  2024-01-19  2024-02-26           9.41         6    56.46   
4       2024-01-05  2024-01-19  2024-02-26           9.41         6    56.46   
...            ...         ...         ...            ...       ...      ...   
2187    2024-12-29  2025-01-04  2025-02-10           9.41         6    56.46   
2188    2024-12-28  2025-01-04  2025-02-10           9.35         5    46.75   
2189    2024-12-28  2025-01-04  2025-02-10           9.41         5    47.05   
2190    2024-12-31  2025-01-04  2025-02-10           9.41         6    56.46   
2191    2024-12-27  2025-01-04  2025-02-10           9.35         5    46.75   

      Classification  
0                  1  
1                  1  
2                  1  
3                  1  
4                  1  
...              ...  
2187               1  
2188               1  
2189               1  
2190               1  
2191               1  

[2192 rows x 16 columns]

In [12]:
#Filtered out the rows which have vendornumber as 4466 in purchas_price
purchases_price=pd.read_sql_query("select * from purchase_prices where vendornumber=4466",engine)
purchases_price

Brand                   Description  Price    Size Volume  Classification  \
0   5215  TGI Fridays Long Island Iced  12.99  1750mL   1750               1   
1   5255  TGI Fridays Ultimte Mudslide  12.99  1750mL   1750               1   
2   3140      TGI Fridays Orange Dream  14.99  1750mL   1750               1   

   PurchasePrice  VendorNumber                   VendorName  
0           9.41          4466  AMERICAN VINTAGE BEVERAGE    
1           9.35          4466  AMERICAN VINTAGE BEVERAGE    
2          11.19          4466  AMERICAN VINTAGE BEVERAGE

In [13]:
sales=pd.read_sql_query("select * from sales where VendorNo=4466",engine)
sales

InventoryId Store Brand                   Description   Size  \
0    1_HARDERSFIELD_5215     1  5215  TGI Fridays Long Island Iced  1.75L   
1    1_HARDERSFIELD_5215     1  5215  TGI Fridays Long Island Iced  1.75L   
2    1_HARDERSFIELD_5215     1  5215  TGI Fridays Long Island Iced  1.75L   
3    1_HARDERSFIELD_5215     1  5215  TGI Fridays Long Island Iced  1.75L   
4    1_HARDERSFIELD_5215     1  5215  TGI Fridays Long Island Iced  1.75L   
..                   ...   ...   ...                           ...    ...   
426     15_WANBORNE_5215    15  5215  TGI Fridays Long Island Iced  1.75L   
427     15_WANBORNE_5255    15  5255  TGI Fridays Ultimte Mudslide  1.75L   
428     15_WANBORNE_5255    15  5255  TGI Fridays Ultimte Mudslide  1.75L   
429     15_WANBORNE_5255    15  5255  TGI Fridays Ultimte Mudslide  1.75L   
430     15_WANBORNE_5255    15  5255  TGI Fridays Ultimte Mudslide  1.75L   

    SalesQuantity SalesDollars SalesPrice   SalesDate Volume Classification  \
0               1        12.99      12.99  09-01-2024   1750              1   
1               1        12.99      12.99  12-01-2024   1750              1   
2               1        12.99      12.99  15-01-2024   1750              1   
3               1        12.99      12.99  21-01-2024   1750              1   
4               1        12.99      12.99  23-01-2024   1750              1   
..            ...          ...        ...         ...    ...            ...   
426             1        12.99      12.99  23-02-2024   1750              1   
427             1        12.99      12.99  03-02-2024   1750              1   
428             1        12.99      12.99  23-02-2024   1750              1   
429             1        12.99      12.99  24-02-2024   1750              1   
430             2        25.98      12.99  27-02-2024   1750              1   

    ExciseTax VendorNo                   VendorName  
0        1.84     4466  AMERICAN VINTAGE BEVERAGE    
1        1.84     4466  AMERICAN VINTAGE BEVERAGE    
2        1.84     4466  AMERICAN VINTAGE BEVERAGE    
3        1.84     4466  AMERICAN VINTAGE BEVERAGE    
4        1.84     4466  AMERICAN VINTAGE BEVERAGE    
..        ...      ...                          ...  
426      1.84     4466  AMERICAN VINTAGE BEVERAGE    
427      1.84     4466  AMERICAN VINTAGE BEVERAGE    
428      1.84     4466  AMERICAN VINTAGE BEVERAGE    
429      1.84     4466  AMERICAN VINTAGE BEVERAGE    
430      3.67     4466  AMERICAN VINTAGE BEVERAGE    

[431 rows x 14 columns]

In [14]:
purchases

InventoryId  Store  Brand                   Description   Size  \
0        30_CULCHETH_5255     30   5255  TGI Fridays Ultimte Mudslide  1.75L   
1       34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
2     1_HARDERSFIELD_5255      1   5255  TGI Fridays Ultimte Mudslide  1.75L   
3       38_GOULCREST_5215     38   5215  TGI Fridays Long Island Iced  1.75L   
4     59_CLAETHORPES_5215     59   5215  TGI Fridays Long Island Iced  1.75L   
...                   ...    ...    ...                           ...    ...   
2187     81_PEMBROKE_5215     81   5215  TGI Fridays Long Island Iced  1.75L   
2188   62_KILMARNOCK_5255     62   5255  TGI Fridays Ultimte Mudslide  1.75L   
2189    34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
2190     6_GOULCREST_5215      6   5215  TGI Fridays Long Island Iced  1.75L   
2191    35_HALIVAARA_5255     35   5255  TGI Fridays Ultimte Mudslide  1.75L   

      VendorNumber                   VendorName  PONumber      PODate  \
0             4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
1             4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
2             4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
3             4466  AMERICAN VINTAGE BEVERAGE        8207  2023-12-27   
4             4466  AMERICAN VINTAGE BEVERAGE        8207  2023-12-27   
...            ...                          ...       ...         ...   
2187          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2188          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2189          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2190          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2191          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   

     ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0       2024-01-01  2024-01-07  2024-02-21           9.35         4    37.40   
1       2024-01-02  2024-01-07  2024-02-21           9.41         5    47.05   
2       2024-01-01  2024-01-07  2024-02-21           9.35         6    56.10   
3       2024-01-07  2024-01-19  2024-02-26           9.41         6    56.46   
4       2024-01-05  2024-01-19  2024-02-26           9.41         6    56.46   
...            ...         ...         ...            ...       ...      ...   
2187    2024-12-29  2025-01-04  2025-02-10           9.41         6    56.46   
2188    2024-12-28  2025-01-04  2025-02-10           9.35         5    46.75   
2189    2024-12-28  2025-01-04  2025-02-10           9.41         5    47.05   
2190    2024-12-31  2025-01-04  2025-02-10           9.41         6    56.46   
2191    2024-12-27  2025-01-04  2025-02-10           9.35         5    46.75   

      Classification  
0                  1  
1                  1  
2                  1  
3                  1  
4                  1  
...              ...  
2187               1  
2188               1  
2189               1  
2190               1  
2191               1  

[2192 rows x 16 columns]

In [15]:
purchases.groupby(['Brand','PurchasePrice'])[['Quantity','Dollars']].sum()

Quantity   Dollars
Brand PurchasePrice                    
3140  11.19              4640  51921.60
5215  9.41               4923  46325.43
5255  9.35               6215  58110.25

In [16]:
#WE CAME TO KNOW THAT PURCHASE TABLE IS MADE WITH THE COMBINATION OF PURCHASE_PRICE AND VENDOR_INVOICE(PO NUMBER IS UNIQUE)

In [17]:
sales['SalesDollars'] = pd.to_numeric(sales['SalesDollars'], errors='coerce')
sales['SalesPrice'] = pd.to_numeric(sales['SalesPrice'], errors='coerce')
sales['SalesQuantity'] = pd.to_numeric(sales['SalesQuantity'], errors='coerce')

sales.groupby(['Brand'])[['SalesDollars','SalesPrice','SalesQuantity']].sum()

SalesDollars  SalesPrice  SalesQuantity
Brand                                         
5215        3897.00     2416.14            300
5255        4273.71     3182.55            329

=>The purchases table contains actual purchase data, including the date of purchase, products (brands) purchased by vendors, the amount paid (in dollars), and the quantity purchased.

=>The purchase price column is derived from the purchase_prices table, which provides product-wise actual and purchase prices. The combination of vendor and brand is unique in this table.

=>The vendor_invoice table aggregates data from the purchases table, summarizing quantity and dollar amounts, along with an additional column for freight. This table maintains uniqueness based on vendor and PO number.

=>The sales table captures actual sales transactions, detailing the brands purchased by vendors, the quantity sold, the selling price, and the revenue earned.

=>As the data that we need for analysis is distributed in different tables, we need to create a summary table containing:

1.purchase transactions made by vendors
2.sales transaction data
3.freight costs for each vendor
4.actual product prices from vendors



Let's first explore each summary in different tables:




In [20]:
vendor_invoice.columns

NameError: name 'vendor_invoice' is not defined

In [ ]:
freight_summary=pd.read_sql_query("select VendorNumber,sum(Freight) as Freight_cost from vendor_invoice group by VendorNumber ",engine)
freight_summary

In [23]:
purchases_price.columns

Index(['Brand', 'Description', 'Price', 'Size', 'Volume', 'Classification',
       'PurchasePrice', 'VendorNumber', 'VendorName'],
      dtype='object')

In [25]:
purchases.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'VendorNumber',
       'VendorName', 'PONumber', 'PODate', 'ReceivingDate', 'InvoiceDate',
       'PayDate', 'PurchasePrice', 'Quantity', 'Dollars', 'Classification'],
      dtype='object')

In [27]:
pd.read_sql_query("""
SELECT 
    p.VendorNumber,
    p.VendorName,
    p.Brand,
    p.PurchasePrice,
    SUM(p.Quantity) AS TotalPurchasedQuantity,
    SUM(p.Dollars) AS TotalPurchasedDollars,
    a.Volume,
    a.Price AS ActualPrice
FROM purchases p
JOIN purchase_prices a ON a.Brand = p.Brand
GROUP BY p.VendorNumber, p.VendorName, p.Brand, p.PurchasePrice, a.Volume, a.Price
ORDER BY TotalPurchasedDollars
""", engine)

VendorNumber                   VendorName  Brand  PurchasePrice  \
0              2561  EDRINGTON AMERICAS            2166           0.00   
1              7245  PROXIMO SPIRITS INC.          3065           0.71   
2              3960  DIAGEO NORTH AMERICA INC      6127           1.47   
3              3924  HEAVEN HILL DISTILLERIES      9123           0.74   
4              8004  SAZERAC CO INC                5683           0.39   
...             ...                          ...    ...            ...   
10688          3960  DIAGEO NORTH AMERICA INC      3545          21.89   
10689          3960  DIAGEO NORTH AMERICA INC      4261          16.17   
10690         17035  PERNOD RICARD USA             8068          18.24   
10691          4425        MARTIGNETTI COMPANIES   3405          23.19   
10692          1128  BROWN-FORMAN CORP             1233          26.27   

       TotalPurchasedQuantity  TotalPurchasedDollars Volume  ActualPrice  
0                      2015.0                   0.00    750         0.00  
1                         1.0                   0.71     50         0.99  
2                         1.0                   1.47    200         1.99  
3                         2.0                   1.48     50         0.99  
4                         6.0                   2.34     50         0.49  
...                       ...                    ...    ...          ...  
10688                138109.0             3023206.01   1750        29.99  
10689                201682.0             3261197.94   1750        22.99  
10690                187407.0             3418303.68   1750        24.99  
10691                164038.0             3804041.22   1750        28.99  
10692                145080.0             3811251.60   1750        36.99  

[10693 rows x 8 columns]

**Here we can see Purchase price of vendor number 2561 is 0.00 which is misspelled or data entry is wrong so we will remove it**

In [ ]:
pd.read_sql_query("""
SELECT 
    p.VendorNumber,
    p.VendorName,
    p.Brand,
    p.PurchasePrice,
   
    a.Volume,
    a.Price AS ActualPrice,
     SUM(p.Quantity) AS TotalPurchasedQuantity,
    SUM(p.Dollars) AS TotalPurchasedDollars 
FROM purchases p
JOIN purchase_prices a ON a.Brand = p.Brand
where p.PurchasePrice>0
GROUP BY p.VendorNumber, p.VendorName, p.Brand, p.PurchasePrice, a.Volume, a.Price
ORDER BY TotalPurchasedDollars
""", engine)

In [29]:
pd.read_sql("SHOW TABLES;", engine)

Tables_in_suviddha_supermarket
0                begin_inventory
1                  end_inventory
2                purchase_prices
3                      purchases
4                          sales
5                 vendor_invoice

In [31]:
sales.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'SalesQuantity',
       'SalesDollars', 'SalesPrice', 'SalesDate', 'Volume', 'Classification',
       'ExciseTax', 'VendorNo', 'VendorName'],
      dtype='object')

In [ ]:
pd.read_sql_query("""select VendorNo,Brand,SUM(SalesPrice) AS TotalSalesPrice,
    SUM(SalesDollars) AS TotalSalesDollars,SUM(SalesQuantity) AS TotalSalesQuantity,
    SUM(ExciseTax) AS TotalExciseTax from sales group by VendorNo,Brand order by TotalSalesDollars""",engine)

In [37]:
# merging all the summary table to create an aggregated summary table for further analysis
suviddha_sales_summary = pd.read_sql_query("""WITH FreightSummary AS (
    SELECT 
        VendorNumber, 
        SUM(Freight) AS FreightCost 
    FROM vendor_invoice 
    GROUP BY VendorNumber
), 

PurchaseSummary AS (
    SELECT 
        p.VendorNumber,
        p.VendorName,
        p.Brand,
        p.Description,
        p.PurchasePrice,
        pp.Price AS ActualPrice,
        pp.Volume,
        SUM(p.Quantity) AS TotalPurchaseQuantity,
        SUM(p.Dollars) AS TotalPurchaseDollars
    FROM purchases p
    JOIN purchase_prices pp
        ON p.Brand = pp.Brand
    WHERE p.PurchasePrice > 0
    GROUP BY p.VendorNumber, p.VendorName, p.Brand, p.Description, p.PurchasePrice, pp.Price, pp.Volume
), 

SalesSummary AS (
    SELECT 
        VendorNo,
        Brand,
        SUM(SalesQuantity) AS TotalSalesQuantity,
        SUM(SalesDollars) AS TotalSalesDollars,
        SUM(SalesPrice) AS TotalSalesPrice,
        SUM(ExciseTax) AS TotalExciseTax
    FROM sales
    GROUP BY VendorNo, Brand
) 

SELECT 
    ps.VendorNumber,
    ps.VendorName,
    ps.Brand,
    ps.Description,
    ps.PurchasePrice,
    ps.ActualPrice,
    ps.Volume,
    ps.TotalPurchaseQuantity,
    ps.TotalPurchaseDollars,
    ss.TotalSalesQuantity,
    ss.TotalSalesDollars,
    ss.TotalSalesPrice,
    ss.TotalExciseTax,
    fs.FreightCost
FROM PurchaseSummary ps
LEFT JOIN SalesSummary ss 
    ON ps.VendorNumber = ss.VendorNo 
    AND ps.Brand = ss.Brand
LEFT JOIN FreightSummary fs 
    ON ps.VendorNumber = fs.VendorNumber
ORDER BY ps.TotalPurchaseDollars DESC""",engine)


**Suviddha Supermarket Sales Summary**

In [41]:
suviddha_sales_summary

VendorNumber                   VendorName  Brand  \
0              1128  BROWN-FORMAN CORP             1233   
1              4425        MARTIGNETTI COMPANIES   3405   
2             17035  PERNOD RICARD USA             8068   
3              3960  DIAGEO NORTH AMERICA INC      4261   
4              3960  DIAGEO NORTH AMERICA INC      3545   
...             ...                          ...    ...   
10687          9815  WINE GROUP INC                8527   
10688          8004  SAZERAC CO INC                5683   
10689          3924  HEAVEN HILL DISTILLERIES      9123   
10690          3960  DIAGEO NORTH AMERICA INC      6127   
10691          7245  PROXIMO SPIRITS INC.          3065   

                       Description  PurchasePrice  ActualPrice Volume  \
0          Jack Daniels No 7 Black          26.27        36.99   1750   
1            Tito's Handmade Vodka          23.19        28.99   1750   
2                 Absolut 80 Proof          18.24        24.99   1750   
3           Capt Morgan Spiced Rum          16.17        22.99   1750   
4                  Ketel One Vodka          21.89        29.99   1750   
...                            ...            ...          ...    ...   
10687  Concannon Glen Ellen Wh Zin           1.32         4.99    750   
10688  Dr McGillicuddy's Apple Pie           0.39         0.49     50   
10689              Deep Eddy Vodka           0.74         0.99     50   
10690  The Club Strawbry Margarita           1.47         1.99    200   
10691     Three Olives Grape Vodka           0.71         0.99     50   

       TotalPurchaseQuantity  TotalPurchaseDollars  TotalSalesQuantity  \
0                   145080.0            3811251.60              9578.0   
1                   164038.0            3804041.22              9203.0   
2                   187407.0            3418303.68             11189.0   
3                   201682.0            3261197.94             20226.0   
4                   138109.0            3023206.01             11883.0   
...                      ...                   ...                 ...   
10687                    2.0                  2.64                 3.0   
10688                    6.0                  2.34               128.0   
10689                    2.0                  1.48                 NaN   
10690                    1.0                  1.47                 6.0   
10691                    1.0                  0.71                81.0   

       TotalSalesDollars  TotalSalesPrice  TotalExciseTax  FreightCost  
0              344712.22         64889.97        17598.14     68601.68  
1              275162.97         52289.50        16909.12    144929.24  
2              288135.11         48202.30        20557.97    123780.22  
3              444810.74         43304.31        37163.76    257032.07  
4              357759.17         52774.51        21833.58    257032.07  
...                  ...              ...             ...          ...  
10687               5.97             5.97            0.33     27100.41  
10688              62.72             0.98            6.72     50293.62  
10689                NaN              NaN             NaN     14069.87  
10690              11.94             7.96            1.26    257032.07  
10691              80.19            29.70            4.21     38994.78  

[10692 rows x 14 columns]

# **DATA CLEANING**

In [49]:
suviddha_sales_summary.dtypes

VendorNumber               int64
VendorName                object
Brand                      int64
Description               object
PurchasePrice            float64
ActualPrice              float64
Volume                    object
TotalPurchaseQuantity    float64
TotalPurchaseDollars     float64
TotalSalesQuantity       float64
TotalSalesDollars        float64
TotalSalesPrice          float64
TotalExciseTax           float64
FreightCost              float64
dtype: object

In [51]:
suviddha_sales_summary.isnull().sum()


VendorNumber                0
VendorName                  0
Brand                       0
Description                 0
PurchasePrice               0
ActualPrice                 0
Volume                      0
TotalPurchaseQuantity       0
TotalPurchaseDollars        0
TotalSalesQuantity       3548
TotalSalesDollars        3548
TotalSalesPrice          3548
TotalExciseTax           3548
FreightCost                 0
dtype: int64

In [53]:
vendor_sales_summary['Volume'].unique()


array(['1750', '750', '1000', '1500', '375', '50', '3000', '5000', '100',
       '200', '4000', '187', '500', '600', '300', '1100', '250', '400',
       '18000', '150', '720', '330', '162.5', '180', '19500', '6000',
       '560', '20000', '9000', '3750', '650'], dtype=object)

In [55]:
vendor_sales_summary['VendorName'].unique()

array(['BROWN-FORMAN CORP          ', 'MARTIGNETTI COMPANIES',
       'PERNOD RICARD USA          ', 'DIAGEO NORTH AMERICA INC   ',
       'BACARDI USA INC            ', 'JIM BEAM BRANDS COMPANY    ',
       'MAJESTIC FINE WINES        ', 'ULTRA BEVERAGE COMPANY LLP ',
       'STOLI GROUP,(USA) LLC      ', 'PROXIMO SPIRITS INC.       ',
       'MOET HENNESSY USA INC      ', 'CAMPARI AMERICA            ',
       'SAZERAC CO INC             ', 'CONSTELLATION BRANDS INC   ',
       'M S WALKER INC             ', 'SAZERAC NORTH AMERICA INC. ',
       'PALM BAY INTERNATIONAL INC ', 'REMY COINTREAU USA INC     ',
       'SIDNEY FRANK IMPORTING CO  ', 'E & J GALLO WINERY         ',
       'WILLIAM GRANT & SONS INC   ', 'HEAVEN HILL DISTILLERIES   ',
       'DISARONNO INTERNATIONAL LLC', 'EDRINGTON AMERICAS         ',
       'CASTLE BRANDS CORP.        ', 'SOUTHERN WINE & SPIRITS NE ',
       'STE MICHELLE WINE ESTATES  ', 'TRINCHERO FAMILY ESTATES   ',
       'MHW LTD                    ', 'W

=>The Volume is a numerical column but having object datatype.
    
=>There are some products that are not sold, so having missing values.
    
=>There are white spaces in the categorical columns.

**Creating new columns for better analysis**

In [59]:
# creating new columns for better analysis
vendor_sales_summary['GrossProfit'] = vendor_sales_summary['TotalSalesDollars'] - vendor_sales_summary['TotalPurchaseDollars']
vendor_sales_summary['ProfitMargin'] = (vendor_sales_summary['GrossProfit'] / vendor_sales_summary['TotalSalesDollars'])*100
vendor_sales_summary['StockTurnover'] = vendor_sales_summary['TotalSalesQuantity'] / vendor_sales_summary['TotalPurchaseQuantity']
vendor_sales_summary['SalesToPurchaseRatio'] = vendor_sales_summary['TotalSalesDollars'] / vendor_sales_summary['TotalPurchaseDollars']

In [61]:
# changing datatype to float
vendor_sales_summary['Volume'] = vendor_sales_summary['Volume'].astype('float')

# filling missing value with 0
vendor_sales_summary.fillna(0,inplace = True)

# removing spaces from categorical columns
vendor_sales_summary['VendorName'] = vendor_sales_summary['VendorName'].str.strip()
vendor_sales_summary['Description'] = vendor_sales_summary['Description'].str.strip()

**Saving cleaned data of suviddha_summary into database**

In [67]:
cursor = conn.cursor()

In [69]:
# only runs once 
cursor.execute("""CREATE TABLE vendor_sales_summary (
    VendorNumber INT,
    VendorName VARCHAR(100),
    Brand INT,
    Description VARCHAR(100),
    PurchasePrice DECIMAL(10,2),
    ActualPrice DECIMAL(10,2),
    Volume INT,
    TotalPurchaseQuantity INT,
    TotalPurchaseDollars DECIMAL(15,2),
    TotalSalesQuantity INT,
    TotalSalesDollars DECIMAL(15,2),
    TotalSalesPrice DECIMAL(15,2),
    TotalExciseTax DECIMAL(15,2),
    FreightCost DECIMAL(15,2),
    GrossProfit DECIMAL(15,2),
    ProfitMargin DECIMAL(15,2),
    StockTurnover DECIMAL(15,2),
    SalesToPurchaseRatio DECIMAL(15,2),
    PRIMARY KEY (VendorNumber, Brand)
);
""")

In [73]:
vendor_sales_summary.to_sql('vendor_sales_summary', engine, if_exists = 'replace', index = False )

10692

# **Suviddha Supermarket Sales Summary Table**

In [77]:
pd.read_sql_query("select * from vendor_sales_summary",engine)

VendorNumber                VendorName  Brand  \
0              1128         BROWN-FORMAN CORP   1233   
1              4425     MARTIGNETTI COMPANIES   3405   
2             17035         PERNOD RICARD USA   8068   
3              3960  DIAGEO NORTH AMERICA INC   4261   
4              3960  DIAGEO NORTH AMERICA INC   3545   
...             ...                       ...    ...   
10687          9815            WINE GROUP INC   8527   
10688          8004            SAZERAC CO INC   5683   
10689          3924  HEAVEN HILL DISTILLERIES   9123   
10690          3960  DIAGEO NORTH AMERICA INC   6127   
10691          7245      PROXIMO SPIRITS INC.   3065   

                       Description  PurchasePrice  ActualPrice  Volume  \
0          Jack Daniels No 7 Black          26.27        36.99  1750.0   
1            Tito's Handmade Vodka          23.19        28.99  1750.0   
2                 Absolut 80 Proof          18.24        24.99  1750.0   
3           Capt Morgan Spiced Rum          16.17        22.99  1750.0   
4                  Ketel One Vodka          21.89        29.99  1750.0   
...                            ...            ...          ...     ...   
10687  Concannon Glen Ellen Wh Zin           1.32         4.99   750.0   
10688  Dr McGillicuddy's Apple Pie           0.39         0.49    50.0   
10689              Deep Eddy Vodka           0.74         0.99    50.0   
10690  The Club Strawbry Margarita           1.47         1.99   200.0   
10691     Three Olives Grape Vodka           0.71         0.99    50.0   

       TotalPurchaseQuantity  TotalPurchaseDollars  TotalSalesQuantity  \
0                   145080.0            3811251.60              9578.0   
1                   164038.0            3804041.22              9203.0   
2                   187407.0            3418303.68             11189.0   
3                   201682.0            3261197.94             20226.0   
4                   138109.0            3023206.01             11883.0   
...                      ...                   ...                 ...   
10687                    2.0                  2.64                 3.0   
10688                    6.0                  2.34               128.0   
10689                    2.0                  1.48                 0.0   
10690                    1.0                  1.47                 6.0   
10691                    1.0                  0.71                81.0   

       TotalSalesDollars  TotalSalesPrice  TotalExciseTax  FreightCost  \
0              344712.22         64889.97        17598.14     68601.68   
1              275162.97         52289.50        16909.12    144929.24   
2              288135.11         48202.30        20557.97    123780.22   
3              444810.74         43304.31        37163.76    257032.07   
4              357759.17         52774.51        21833.58    257032.07   
...                  ...              ...             ...          ...   
10687               5.97             5.97            0.33     27100.41   
10688              62.72             0.98            6.72     50293.62   
10689               0.00             0.00            0.00     14069.87   
10690              11.94             7.96            1.26    257032.07   
10691              80.19            29.70            4.21     38994.78   

       GrossProfit  ProfitMargin  StockTurnover  SalesToPurchaseRatio  
0      -3466539.38  -1005.632867       0.066019              0.090446  
1      -3528878.25  -1282.468440       0.056103              0.072334  
2      -3130168.57  -1086.354443       0.059704              0.084292  
3      -2816387.20   -633.165287       0.100287              0.136395  
4      -2665446.84   -745.039419       0.086041              0.118338  
...            ...           ...            ...                   ...  
10687         3.33     55.778894       1.500000              2.261364  
10688        60.38     96.269133      21.333333             26.803419  
10689         0.00 